### ISSUE TITLE
BOSH director can be stalled to unusability by bosh tasks -d -r

### ISSUE THREAD

**Issue by schmidtsv** [cloudfoundry/bosh#2105]
**Describe the bug**
During our operations we found an issue where the bosh director can become unresponsive when bosh tasks -r -d is called with a higher value for -r than tasks are stored in the database. Example:
```
$ time  bosh tasks  -d red537633 -r=1
Using environment '172.16.1.17' as client 'admin'

ID        State  Started At                    Last Activity At              User     Deployment  Description        Result  
17950927  done   Tue Dec 18 15:16:35 UTC 2018  Tue Dec 18 15:20:18 UTC 2018  service  red537633   create deployment  /deployments/red537633  

1 tasks

Succeeded

real	0m0.065s
user	0m0.032s
sys	0m0.004s
[16:08:41] a9s@aws-staging ~
$ time  bosh tasks  -d red537633 -r=2
Using environment '172.16.1.17' as client 'admin'

ID        State  Started At                    Last Activity At              User     Deployment  Description        Result  
17950927  done   Tue Dec 18 15:16:35 UTC 2018  Tue Dec 18 15:20:18 UTC 2018  service  red537633   create deployment  /deployments/red537633  

1 tasks

Succeeded

real	1m46.813s
user	0m0.032s
sys	0m0.000s
```
We have created that deployment freshly, so only one task has been performed on it. When we try to use -r=2, you can see that the time to complete the call takes 1:45. We assume this happens because the Director DB just crawls backwards until it found enough tasks. We currently have 3.5m tasks in the database which means that going through the full dataset can take a very long time. During that time the Director becomes very unresponsive and some tasks can time out or cannot be created due to the Director not responding. We see this issue with the endpoint: `/tasks?deployment=<deployment_name>&limit=30&verbose=1`

**To Reproduce**
Set up a larger BOSH database. Ensure that the queries are not cached so far. Our bosh director has 17m tasks and the tasks table in the database contains. 3.5m tasks. It runs on a m4.large instance with 500gb GP2 volume. I was having a Zoom Meeting with @dpb587 about it where I showed how the bug most likely stems from. When you do a tasks -r it seems it just crawls through the database backwards until it found enough matching tasks. If it does not, it crawls through the entire table. 

**Expected behavior**
Even when I try to get more tasks than exist in the Database for a given deployment, I get them in a timely fashion

**Logs**
These are some sample log lines of requests for the tasks of one deployment:
```
admin [18/Dec/2018:15:06:13 +0000] "GET /tasks?deployment=d52f2d3&limit=30&verbose=1 HTTP/1.1" 499 0 "-" "Go-http-client/1.1" 35.356 - .
admin [18/Dec/2018:15:07:17 +0000] "GET /tasks?deployment=d52f2d3&limit=9&verbose=1 HTTP/1.1" 499 0 "-" "Go-http-client/1.1" 36.524 - .
admin [18/Dec/2018:15:09:18 +0000] "GET /tasks?deployment=d52f2d3&limit=30&verbose=1 HTTP/1.1" 200 1761 "-" "Go-http-client/1.1" 117.158 117.154 .
```
**Versions (please complete the following information):**
 - Infrastructure: AWS
 - BOSH version 268.0.1
 - BOSH CLI version 2.0.45-d208799-2017-10-28T00:31:53Z
 - Stemcell version ubuntu-xenial/"97.12"
 - BPM/ 0.12.3

**Deployment info:**
Deployed via bosh deployment based on commit "4d43709fe57144386532bbd5c839f4a3a919233c". The only change compared to normal bosh deployment is the addition of external UAA and Credhub. Ops files used from bosh-deployment are:
    -o aws/cpi.yml 
    -o local-dns.yml 
    -o misc/dns.yml 

**Additional context**
It might be hard to get those times we get on your own test setup, what could be an idea would be to add an index to deployment_name (so less db content gets crawled) and/or adding an index to the task type to easier filter for specific task types when selecting. Additionally, I want to note that the default number of Puma Workers on the director is 3, which means that you can make this blocking if you have all worker threads blocked by those long running calls. (https://github.com/cloudfoundry/bosh/blob/master/jobs/director/spec#L79).

**Comment by mfine30** [comment]
I believe that we have addressed this issue as it seems very similar to the work we did in the story: [#163150850](https://www.pivotaltracker.com/n/projects/956238/stories/163150850). We have since released a director with these fixes. If you are on a more recent version of director and continue to see this as an issue, please open a new issue and let us know!

Thanks!
